Update getmanifest to create unique filename for local copy of manifest

Multimac opened the pull request:
One thing I noticed in the way Munki handles manifests is that it uses the filename part when storing locally. This means that having multiple manifests with the same name but in different folders will cause clients to play up.

This changes the local naming scheme for manifests to allow multiple manifests with the same name, but in different folders on the server. It does this by including the relative folder portion of the path in the manifest filename.

erikng commented:
I'm curious as to why you are putting your manifests in folders. 

Sent from my iPhone

> On Jun 21, 2015, at 10:08 PM, Multimac <notifications@github.com> wrote:
> 
> One thing I noticed in the way Munki handles manifests is that it uses the filename part when storing locally. This means that having multiple manifests with the same name but in different folders will cause clients to play up.
> 
> This changes the local naming scheme for manifests to allow multiple manifests with the same name, but in different folders on the server. It does this by including the relative folder portion of the path in the manifest filename.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/munki/munki/pull/490
> 
> Commit Summary
> 
> Update getmanifest to create unique filename for local copy of manifest
> File Changes
> 
> M code/client/munkilib/updatecheck.py (5)
> Patch Links:
> 
> https://github.com/munki/munki/pull/490.patch
> https://github.com/munki/munki/pull/490.diff
> —
> Reply to this email directly or view it on GitHub.
>

Multimac commented:
Find it helps keep them a bit more organised.

I've set it up so there's a couple global manifests, followed by a folder per department with department-wide manifests, finally followed by a folder containing each computer's settings.

Have uploaded a screenshot of my folder structure. The printers manifest in '/manifests/' conflicts with the manifest in '/manifests/dept/it/'.

![screen shot 2015-06-22 at 2 22 00 pm](https://cloud.githubusercontent.com/assets/354845/8275504/4c921e6e-18ea-11e5-965f-26811e054074.png)

erikng commented:
I do something similar but I've never hit this issue (actually surprised I haven't). 

Sent from my iPhone

> On Jun 21, 2015, at 11:27 PM, Multimac <notifications@github.com> wrote:
> 
> Find it helps keep them a bit more organised.
> 
> I've set it up so there's a couple global manifests, followed by a folder per department with department-wide manifests, finally followed by a folder containing each computer's settings.
> 
> Have uploaded a screenshot of my folder structure. The printers manifest in '/manifests/' conflicts with the manifest in '/manifests/dept/it/'.
> 
> 
> 
> —
> Reply to this email directly or view it on GitHub.
>

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2572,7 +2572,10 @@ def getmanifest(partialurl, suppress_errors=False):
+        
+        valid_chars = "-_.() %s%s" % (string.ascii_letters, string.digits)
+        manifestname = ''.join(c if c in valid_chars else "_" for c in partialurl)
+
```
Not sure why you are doing this substitution. This could cause the the same problem you are attempting to solve:
manifests/dept1/faculty/printers and manifests/dept1_faculty/printers would be stored as the same name.

Multimac commented:
Yeah, very good point. Thought that may be a problem.

I've tweaked the caching method to mirror the folder structure on the server, so if a manifest is ".../manifests/dept/it/base", it will be in the same location on the client (only in "/Library/Managed Installs/manifests" instead).

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2572,15 +2608,24 @@ def getmanifest(partialurl, suppress_errors=False):
-        return MANIFESTS[manifestname]
+    if getFromManifestCache(manifestname)
+        return getFromManifestCache(manifestname)
 
```
Seems less than optimal to call getFromManifestCache() twice.

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2553,6 +2553,42 @@ def getmanifest(partialurl, suppress_errors=False):
+    def walkManifestCache(path):
+        """Walks MANIFESTS to the specified path, creating any intermediate
+        dictionaries if they don't exist.
+
```
Doc string doesn't match actual behavior here

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2553,6 +2553,42 @@ def getmanifest(partialurl, suppress_errors=False):
+    def addToManifestCache(path, value):
+        """Sets the item at the specified path to the given value.
+        """
+        parts = os.path.split(path);
```
Python lines don't end with a semicolon

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2553,6 +2553,42 @@ def getmanifest(partialurl, suppress_errors=False):
+        Returns:
+            The item at path in MANIFESTS
+        """
+        parts = os.path.split(path);
```
same here

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2572,15 +2608,24 @@ def getmanifest(partialurl, suppress_errors=False):
     munkicommon.display_debug2('Manifest base URL is: %s', manifestbaseurl)
     munkicommon.display_detail('Getting manifest %s...', manifestdisplayname)
     manifestpath = os.path.join(manifest_dir, manifestname)
+    manifestdir = os.path.dirname(manifestpath)
```
It's confusing to have manifest_dir and manifestdir

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2553,6 +2553,42 @@ def getmanifest(partialurl, suppress_errors=False):
     """
     #global MANIFESTS
+
+    def walkManifestCache(path):
```
Not really sure why we've added these functions. Can't we just continue using the MANIFESTS dictionary, but use the full (or relative) path as the key instead of just the filename?

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2605,7 +2650,7 @@ def getmanifest(partialurl, suppress_errors=False):
+        addToManifestCache(manifestname, manifestpath)
         return manifestpath
 
 
```
Probably could now also use a new cleanUpManifests() function, similar to cleanUpCatalogs, to remove any manifests from the local cache that are no longer used, including any empty subdirectories. (Be sure to leave the SelfServeManifest alone if it exists!)

Multimac commented:
Good point on sub directories in MANIFESTS, was trying to solve a non-existent problem. Have reverted to storing relative paths (to root manifest folder) in MANIFESTS.

Also created the cleanUpManifests function as mentioned (using cleanUpCatalogs as a base). Runs recursively to check sub directories. SelfServeManifest can't be called anything different or located anywhere other than the root manifest directory, can it?

gregneagle commented:
"SelfServeManifest can't be called anything different or located anywhere other than the root manifest directory, can it?" Not without breaking every Munki user who already has a SelfServeManifest...

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2608,6 +2617,39 @@ def getmanifest(partialurl, suppress_errors=False):
+
+    # Remove folder if empty
+    try:
+        os.rmdir(manifest_dir)
```
This would have the effect of removing the "root" manifests directory if it's empty, I think we should avoid that.

gregneagle replied:
Might I suggest https://github.com/munki/munki/blob/master/code/client/munkilib/updatecheck.py#L2825-L2842
as a model for a slightly different implementation?

gregneagle commented on code/client/munkilib/updatecheck.py:
```diff
@@ -2581,6 +2581,15 @@ def getmanifest(partialurl, suppress_errors=False):
+        os.makedirs(destinationdir)
+    except OSError:
+        if not os.path.isdir(destinationdir):
+            raise
```
Why bother with the except if we are only going to raise? Either use munkicommon.display_error() to give a more helpful error message, use ManifestException to do something similar, or don't bother at all.  The exception will be raised anyway...

Multimac commented:
Oh, that question about the SelfServeManifest was just making sure it couldn't be placed anywhere else, in case people could do tricky things.

Good point on cleanUpManifests, I've rewritten it to use os.walk instead of recursion.

Also re-raising the os.makedirs exception is to filter out the case where the directory already exists. I've changed it to use display_error() and return None instead of re-raising the exception.
